CorsixTH/Lua/app.lua

```diff
@@ -1531,15 +1537,48 @@ end
 
 --! Function to check the loaded game is compatible with the program
 --!param save_version (num)
+--!param gfx_set (string) What graphics set is used
 --!return true if compatible, otherwise false
-function App:checkCompatibility(save_version)
+function App:checkCompatibility(save_version, gfx_set)
   local app_version = self.savegame_version
-  if app_version >= save_version or self.config.debug then
+  local err
+
+  -- First check the graphics set matches with the game files
+  if (gfx_set == "demo" and not self.using_demo_files) then
+    err = _S.errors.compatibility_error.demo_in_full
+  elseif (gfx_set == "full" and self.using_demo_files) then
+    err = _S.errors.compatibility_error.full_in_demo
+
+  -- if that's all good, check the save and app version
+  elseif app_version >= save_version or self.config.debug then
     return true
-  else
-    UILoadGame:loadError(_S.errors.compatibility_error)
-    return false
+  else -- savegame newer than application
+    err = _S.errors.compatibility_error.new_in_old
+  end
+
+  UILoadGame:loadError(err)
+  return false
+end
+
+--! Compatibility function to work out the game's graphics set
+--!param state The savegame data
+function App.configureGraphicsSetFlag(state)
+  -- First eliminate all but the first campaign level (same in demo and full)
+  local self = state
+  local toxicity = tostring(_S.level_names[1]:upper())
+  local level_name = tostring(self.map.level_name)
+  if level_name ~= toxicity then
+    self.world.gfx_set = "full"
```

Comment:
Why so many tests, if this fails it doesn't matter - remove the level name test if baldness is sufficient.

Reply:
Sadly I don't think checking just baldness is enough.
Other levels in the game use baldness, and that could be falsely marking levels if we don't hone in on a specific level.
Additionally, the demo is called "Demo Town" in the full game and includes baldness.

Reply:
Then is the first check on the level name enough?

Reply:
No. The demo level in the demo, and the first level in the main campaign are both called "ToxiCity" (or the name in the respective original languages).